_posts/2017-11-27-Babel-7-and-TypeScript.markdown

```diff
@@ -0,0 +1,261 @@
+---
+
+At Artsy we <3 [TypeScript](https://www.typescriptlang.org/). We use it with React Native via [Emission](https://github.com/artsy/emission) and on the web via [Reaction](https://github.com/artsy/reaction). Until recently, however, existing projects built on top of Babel had to implement convoluted tooling pipelines in order to work with the TypeScript compiler, increasing friction in an already complex landscape. Thankfully, those days [are over](https://github.com/babel/babel/tree/master/packages/babel-preset-typescript). Read on for an example project, as well as some advice on how to avoid common pitfalls when working with the new beta version of Babel 7.
+
+<!-- more -->
+
+Babel configurations can be complicated. They take time to set up and maintain and can often contain some pretty [far-out features](https://github.com/kentcdodds/babel-macros) that make interop with other environments difficult. That's why we were elated when [this PR](https://github.com/babel/babylon/pull/523) appeared in the wild from [@andy-ms](https://github.com/andy-ms), a developer on the TypeScript team, announcing a new parser for Babylon. [@babel/preset-typescript](https://github.com/babel/babel/tree/master/packages/babel-preset-typescript) arrived soon after and we felt it was finally time to give it a try. There was a catch, however: TypeScript support only works with Babel 7+!
+
+**TLDR; Check out https://github.com/damassi/babel-7-typescript-example**
+
+Here's list of setup issues we faced in no specific order:
+
+### 1) New @babel Namespace
+
+One of the first things Babel 7 users will notice is the package ecosystem now exists as a [monorepo](https://github.com/babel/babel/tree/master/packages) and all NPM modules are namespaced behind the `@babel` org address. Packages that used to be installed via
+
+```sh
+yarn add -D \
+  babel-core \
+  babel-preset-react \
+  babel-preset-stage-3
+  ...
+```
+are now installed via
+```sh
+yarn add -D \
+  @babel/core \
+  @babel/preset-react \
+  @babel/preset-stage-3
+  ...
+```
+which immediately creates upgrade conflicts between libraries that use Babel 6 and Babel 7. For example, `babel-jest` internally points to `babel-core` which supports a [version range between 6 and 7](https://github.com/facebook/jest/blob/master/packages/babel-jest/package.json#L19) -- but! -- `babel-core` is now `@babel/core` so this breaks.
+
+This wasn't immediately apparent at the time, and so we would often find errors like
+
+```sh
+Error: Could not find preset "@babel/env" relative to directory
+```
+
+These errors appeared ambiguous because the folder structure was correct and commands like `yarn list @babel/preset-env` yielded expected results:
+
+```sh
+└─ @babel/preset-env@7.0.0-beta.32
+✨  Done in 0.58s.
+```
+
+Why was the package not found? Digging deeper, it seemed like Babel 6 was still being used somewhere. Running `yarn list babel-core` revealed the culprit:
+
+```sh
+└─ babel-core@6.25.0
+✨  Done in 0.58s.
+```
```

Comment:
Is this the correct snippet? I was expecting to see it show how Babel 6 was still a transitive dependency.

Reply:
Yup, its correct - `babel-core`, rather than `@babel/core` and with a version number of `6.25.0`.